rfcs/20230418-constant-tensor-cache-capacity-control/README.md

```diff
@@ -0,0 +1,226 @@
+Graph interface layer. Each backend should use this interface to cache all their
+constant tensors in to the unified cache. And the oneDNN Graph API users could
+control the capacity of the unified cache, no need to be aware of any backend
+specific things. This proposal is based on the following considerations:
+
+- In memory limited scenarios, we usually want to control the total memory
+  footprint, instead of the memory footprint of specific backend.
+- By the design of oneDNN Graph API, the API users don't need to be aware about
+  which backend is using. We don't like to break this design currently.
+- Backend selection is dynamic in mixed backend scenarios. Users can't
+  explicitly know which backend will run and how many partition will a backend
+  claims until graph partition stage done. So, it's hard to decide how many
+  memory a backend should use.
+
+The key and value design of unified constant tensor cache utils:
+
+- Cache key: The unified cache key can be composed of two parts: backend specific
+  key and backend id. The backend id is unique among backends. The backend
+  specific key can be defined in each backends, to achieve more flexibility.
+- Cache value: The unified cache value can be a base buffer class like following,
+  and each backend can inherit this base class and subclass it.
+
+```cpp
+class constant_buffer_t{
+public:
+    virtual size_t get_size() const;
+    virtual ~constant_buffer_t();
+
+private:
+    void *data_;
+    size_t size_;
+    // Memory freeing relies on engine and allocator
+    const dnnl::engine eng_;
+    const allocator_t *alc_;
+};
+```
+
+The constant tensor cache utils may looks like the following. It will be
+templated class which can be specialized and instantiated for each engine kind
+for dedication purpose. The support for concurrent execution is same as the
+existing DNNL backend constant tensor cache, which used the promise-future
+mechanism. That is not changed.
+
+```cpp
+template<class engine_kind>
+class unified_constant_tensor_cache_t {
+public:
+    using key_t = size_t;
+    using cached_t = std::shared_ptr<constant_buffer_t>;
+    using value_t = std::shared_future<cached_t>;
+    using promise_t = std::promise<cached_t>;
+
+    unified_constant_tensor_cache_t();
+    ~unified_constant_tensor_cache_t();
+
+    status_t set_capacity(size_t capacity);
+    size_t get_capacity();
+
+    /// @backend_id:
+    /// @backend_specific_key:
+    /// @size: The size the constant tensor
+    /// @value: A promised future constant tensor. If cache miss, the thread
+    /// which firstly query the constant tensor will promise to compute it in
+    /// the future, then other threads who also query the same constant tensor
+    /// will wait util the first thread ready.
+    /// @return: A valid future of cache hit, and an invalid one if cache miss.
+    value_t get_or_add(size_t backend_id, size_t backend_specific_key,
```

Comment:
Is this API thread safe?

Reply:
Yes, it's thread safe. Please refer to the description about concurrent execution in the Introduction section

Reply:
The description is not clear to me: `If cache miss, the thread which firstly query the constant tensor will promise to compute it in the future, then other threads who also query the same constant tensor will wait util the first thread ready.`

How to ensure that? What if the first thread doesn't want to provide the cache when cache is missing? For example, when capacity is full, the first thread decides not to add this constant tensor into cache.

Reply:
when the first thread add this constant tensor into cache, other threads will cache hit. when the first thread not add, others threads will cache miss and compute the tensor by themselves.

I will add more descriptions about the cache behaviors in multithreading